src/web/middlewares/web_middleware_csrf.ml

```diff
@@ -1,6 +1,39 @@
-(* TODO
-https://docs.djangoproject.com/en/3.0/ref/csrf/#how-it-works *)
+let ( let* ) = Lwt.bind
 
-let m () =
-  let filter handler ctx = handler ctx in
-  Web_middleware_core.create ~name:"csrf" filter
+module CsrfService = struct
+  let add_to_ctx ctx ~token () =
+    let open Core_ctx in
+    let key : Token.t key = create_key () in
```

Comment:
@aronerben This might be a severe bug. You have to create the key on the top level (when the module gets initialized) since it is a unique global key. If you drop the key like this after adding to the ctx, you can not retrieve the content anymore.

Reply:
I don't understand this. What do you mean by dropping the key? Does `add` mutate the `ctx`?

Reply:
Discussed in person: Key still has to be referenced in global scope, so instead of creating a new one, reference the globally defined one.